### PR TITLE
[AutoDiff] Correctly propagate optional adjoint through `switch_enum`

### LINKED ISSUE #74978: Incorrect gradients in some optional-related cases
### Description

The code below should produce identical gradients, however, it does not

### Reproduction

```swift
import _Differentiation

@differentiable(reverse)
func square(y: Double) -> Double? {
  return y * y
}

@differentiable(reverse)
func square2(y: Double?) -> Double {
  return y! * y!
}

print(valueWithGradient(at: 42.0, of: {y in square(y:y)!}))
print(valueWithGradient(at: 42.0, of: {y in square2(y:y)}))
```

### Expected behavior

(value: 1764.0, gradient: 84.0)
(value: 1764.0, gradient: 84.0)

### Environment

Swift version 6.0-dev (LLVM 76ba61e88550d3b, Swift 9470ad7be9ab5e1)
Target: arm64-apple-macosx13.0

### Additional information

Currently we are printing:
(value: 1764.0, gradient: 84.0)
(value: 1764.0, gradient: 42.0)

### PR DESCRIPTION
**Pull request by asl**
Fixes #74978

### CODE REVIEW

**Comment by asl** [comment]
@swift-ci please test

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp [hunk 1]
```diff
@@ -2734,7 +2735,7 @@ void PullbackCloner::Implementation::accumulateAdjointForOptionalBuffer(
 }
 
 // Set the adjoint value for the incoming `Optional` value.
```
Comment (rxwei) [inline lib/SILOptimizer/Differentiation/PullbackCloner.cpp:2737]: Update the comment?

**Comment by asl** [comment]
So, the code in question looks like this:
```llvm
// square2(y:)
sil hidden [ossa] @$s3opt7square21yS2dSg_tF : $@convention(thin) (Optional<Double>) -> Double {
// %0 "y"                                         // users: %14, %3, %1
bb0(%0 : $Optional<Double>):
  debug_value %0 : $Optional<Double>, let, name "y", argno 1 // id: %1
  %2 = metatype $@thin Double.Type                // user: %26
  switch_enum %0 : $Optional<Double>, case #Optional.some!enumelt: bb2, case #Optional.none!enumelt: bb1 // id: %3
...
// %13                                            // user: %26
bb2(%13 : $Double):                               // Preds: bb0
  switch_enum %0 : $Optional<Double>, case #Optional.some!enumelt: bb4, case #Optional.none!enumelt: bb3 // id: %14
...
// %24                                            // user: %26
bb4(%24 : $Double):                               // Preds: bb2
  // function_ref static Double.* infix(_:_:)
  %25 = function_ref @$sSd1moiyS2d_SdtFZ : $@convention(method) (Double, Double, @thin Double.Type) -> Double // user: %26
  %26 = apply %25(%13, %24, %2) : $@convention(method) (Double, Double, @thin Double.Type) -> Double // user: %27
  return %26 : $Double                            // id: %27
} // end sil function '$s3opt7square21yS2dSg_tF'
```

However, instead of adjoint accumulation we just created new adjoint each time. Funny enough, adjoint buffer version for optionals of non-loadable types was correct – adjoint buffers were accumulated together.

Likely the non-options `switch_enum` case is broken as well, I just do not have a testcase that directly exposes the problem.

**Review by rxwei — APPROVED** [review]
Nice catch!

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp (hunk 1 above)
Comment (asl) [reply lib/SILOptimizer/Differentiation/PullbackCloner.cpp]: Yes, forgot to update here.

**Comment by asl** [comment]
@swift-ci please test

**Comment by asl** [comment]
@swift-ci please test

**Pull request merged**